Registry change: what you need to know blog article

fsmunoz opened the pull request:
Blog article about the registry change impacts and how to address them.

(I'll add more when I have the emails, they're hidden):

Co-authored-by: Davanum Srinivas <davanum@gmail.com>

sftim commented:
/hold

OK to unhold once this has LGTMs from a couple of contributors and the date looks right (it needs to be in the future at the point the article merges)

sftim commented on content/en/blog/_posts/2023-03-10-registry-change-what-you-need-to-know.md:
```diff
@@ -0,0 +1,6 @@
+---
+layout: blog
+title: "k8s.gcr.io redirect to registry.k8s.io - what you need to know"
+date: 2023-3-10
```
```suggestion
date: 2023-03-10
```

More idiomatic.
Also please consider adding a time with timezone.

sftim reviewed (commented):
Let's give this a short URL (see dockershim entries in `static/_redirects` for an example)

I suggest: https://k8s.io/image-registry-change

fsmunoz commented:
@sftim I will update the publication date (and file name) since I don't think this will be published today. Will add a timezone then!

fsmunoz commented:
Updated with the result of the joint work, fixed date format, added redirect.

cc Comms (@chris-short @AvineshTripathi), + @mrbobbytables @dims @katcosgrove @nimbinatus @sftim @gracenng

fsmunoz commented:
/retitle Registry change: what you need to know blog article

chris-short commented on content/en/blog/_posts/2023-03-10-image-registry-change.md:
```diff
@@ -1,7 +1,7 @@
 layout: blog
 title: "k8s.gcr.io redirect to registry.k8s.io - what you need to know"
-date: 2023-3-10T00:00:00.000Z
+date: 2023-03-10T00:00:00.000Z
```
1700Z at the latest, please.

sftim commented on content/en/blog/_posts/2023-03-10-image-registry-change.md:
```diff
@@ -0,0 +1,190 @@
+GCR at the same time.
+
+For more details on this change, check out [registry.k8s.io: faster, cheaper and Generally Available
+(GA)](https://kubernetes.io/blog/2022/11/28/registry-k8s-io-faster-cheaper-ga/).
```
```suggestion
(GA)](/blog/2022/11/28/registry-k8s-io-faster-cheaper-ga/).
```

I might skip the “(GA)” there!

sftim commented on content/en/blog/_posts/2023-03-10-image-registry-change.md:
```diff
@@ -0,0 +1,190 @@
+
+If you believe you have encountered a bug with the new registry or the redirect, please open an
+issue in the [kubernetes/registry.k8s.io
+repo](https://github.com/kubernetes/registry.k8s.io). **Please check if there is an issue already
```
```suggestion
repo](https://github.com/kubernetes/registry.k8s.io/issues/new/choose). **Please check if there is an issue already
```

This lets us add in a “are you asking for support? Well, go here” page in the future, if we want to.

sftim commented on content/en/blog/_posts/2023-03-10-image-registry-change.md:
```diff
@@ -0,0 +1,190 @@
+2023](https://kubernetes.io/blog/2023/02/06/k8s-gcr-io-freeze-announcement/).
+
+Information on the architecture of registry.k8s.io and its [request handling decision
+tree](https://github.com/kubernetes/registry.k8s.io/blob/main/cmd/archeio/docs/request-handling.md)
```
```suggestion
tree](https://github.com/kubernetes/registry.k8s.io/blob/8408d0501a88b3d2531ff54b14eeb0e3c900a4f3/cmd/archeio/docs/request-handling.md)
```

Permalinking in case that file ever moves

sftim commented on content/en/blog/_posts/2023-03-10-image-registry-change.md:
```diff
@@ -0,0 +1,190 @@
+
+If you would like to know more about the image freeze and the last images that will be available
+there, see the blog post: [k8s.gcr.io Image Registry Will Be Frozen From the 3rd of April
+2023](https://kubernetes.io/blog/2023/02/06/k8s-gcr-io-freeze-announcement/).
```
```suggestion
2023](/blog/2023/02/06/k8s-gcr-io-freeze-announcement/).
```

sftim commented on content/en/blog/_posts/2023-03-10-image-registry-change.md:
```diff
@@ -0,0 +1,190 @@
+
+## I still have questions, where should I go?
+
+For more information on registry.k8s.io and why it was developed, see [registry.k8s.io: faster,
+cheaper and Generally Available
+(GA)](https://kubernetes.io/blog/2022/11/28/registry-k8s-io-faster-cheaper-ga/).
```
```suggestion
For more information on registry.k8s.io and why it was developed, see [registry.k8s.io: faster,
cheaper and Generally Available](/blog/2022/11/28/registry-k8s-io-faster-cheaper-ga/).
```

sftim commented on content/en/blog/_posts/2023-03-10-image-registry-change.md:
```diff
@@ -0,0 +1,190 @@
+**Option 5**: As a **LAST** possible option, you can use a [Mutating Admission
+Webhook](https://github.com/slackhq/simple-kubernetes-webhook/blob/main/dev/gen-certs.sh) to change
+the image address dynamically. This should only be considered a stopgap till your manifests have
+been updated. An example Mutating Webhook and Kyverno policy can be found
+[here](https://github.com/abstractinfrastructure/k8s-gcr-quickfix).
```
```suggestion
been updated. You can find a (third party) Mutating Webhook and Kyverno policy in
[k8s-gcr-quickfix](https://github.com/abstractinfrastructure/k8s-gcr-quickfix).
```

chris-short commented on content/en/blog/_posts/2023-03-10-image-registry-change.md:
```diff
@@ -0,0 +1,190 @@
+
+k8s.gcr.io is hosted on a custom [Google Container Registry
+(GCR)](https://cloud.google.com/container-registry) domain that was set up solely for the Kubernetes
+project.This has worked well since the inception of the project, and we thank Google for providing
```
"project.This"

This needs one space

sftim commented on content/en/blog/_posts/2023-03-10-image-registry-change.md:
```diff
@@ -0,0 +1,190 @@
+that will block them from being pulled. You can use these third-party policies with any Kubernetes
+cluster.
+
+**Option 5**: As a **LAST** possible option, you can use a [Mutating Admission
+Webhook](https://github.com/slackhq/simple-kubernetes-webhook/blob/main/dev/gen-certs.sh) to change
+the image address dynamically. This should only be considered a stopgap till your manifests have
```
```suggestion
**Option 5**: As a **LAST** possible option, you can use a Mutating Admission
Webhook to change the the container image dynamically, at Pod creation time
```

https://github.com/slackhq/simple-kubernetes-webhook/blob/main/dev/gen-certs.sh is a shell script and doesn't seem to be the right thing to link to. Could link to https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#what-are-admission-webhooks though.

sftim commented on content/en/blog/_posts/2023-03-10-image-registry-change.md:
````diff
@@ -0,0 +1,190 @@
+FailedCreatePodSandBox: Failed to create pod sandbox: rpc error: code = Unknown desc = Error response from daemon: Head “https://us-west1-docker.pkg.dev/v2/k8s-artifacts-prod/images/pause/manifests/3.8”: x509: certificate signed by unknown authority
+```
+
+## How can I find which images are using the legacy registry?
````
```suggestion
## How can I find which images are using the legacy registry, and fix them?
```

Some of the options are more about the fixing than the finding.

chris-short commented on content/en/blog/_posts/2023-03-10-image-registry-change.md:
```diff
@@ -0,0 +1,190 @@
+  to the new registry. 
+- A small subset of non-standard clients do not handle HTTP redirects by image registries, and will
+  need to be pointed directly at registry.k8s.io.
+- The redirect is a stopgap to assist users in making the switch. The deprecated k8s.gcr.io registry
```
"The deprecated k8s.gcr.io registry"

The sunset k8s.gcr.io...

chris-short replied:
I dunno... Sounds weird maybe

chris-short commented:
/lgtm

fsmunoz commented:
Rebased after updating. Good to go for me!

chris-short commented:
/lgtm

chris-short commented:
/I hold

chris-short commented:
/untold

chris-short commented:
/unhold

dims commented:
/approve
/lgtm

fsmunoz commented:
/assign krol3

mrbobbytables reviewed (commented):
/approve

BenTheElder commented on content/en/blog/_posts/2023-03-10-image-registry-change.md:
```diff
@@ -0,0 +1,190 @@
+
+## What will happen to k8s.gcr.io?
+
+Separate from the the redirect, k8s.gcr.io will be frozen [and will not be updated with new images
```
I don't know how much we want to go into here, but actually for anyone hitting the redirect to registry.k8s.io instead of the existing backing GCRs, all new images will be available at k8s.gcr.io by way of the redirect.

Somewhat confusing situation

dims replied:
@BenTheElder i wanted to leave it vague... as they just need to know that they need to get off!

reylejano commented on content/en/blog/_posts/2023-03-10-image-registry-change.md:
```diff
@@ -0,0 +1,190 @@
+  to the new registry. 
+- A small subset of non-standard clients do not handle HTTP redirects by image registries, and will
+  need to be pointed directly at registry.k8s.io.
+- The redirect is a stopgap to assist users in making the switch. The deprecated k8s.gcr.io registry
```
imo, sounds like the date is moved forward if there is a redirect from k8s.gcr.io starts on March 20

reylejano commented:
There are several lgtms and approves from contributors
I added a few comments/nits that can be addressed in a future PR
/approve

reylejano commented:
reapplying lgtm after a push for the title's case change
/lgtm

thockin commented on content/en/blog/_posts/2023-03-10-image-registry-change.md:
````diff
@@ -0,0 +1,190 @@
+When you run the command above, here’s what to expect when things work correctly:
+
+```
+$ kubectl run hello-world --tty --rm -i --image=registry.k8s.io/busybox:latest sh
````
This is a weird example - why force interactivity?  I think something like `date` instead of `sh` is simpler

thockin commented on content/en/blog/_posts/2023-03-10-image-registry-change.md:
```diff
@@ -0,0 +1,190 @@
+
+If you think you may be impacted, or would like to know more about this change, please keep reading.
+
+## Why did Kubernetes change to a different image registry?
```
I would move the background and rationale sections to the end and move the sections on what is impacted, how do I know, and what do I do about it to the top.  In short, assume people will get fatigued and stop reading, make the most actionable parts first.  The history and funding and future of k8s.gcr.io is the least important part of this.

thockin reviewed (commented):
Thanks for this.  It's a little long, for my tastes, but there's not much that should obviously be dropped.

I would suggest moving the "why ..." sections to the end.

thockin commented:
I offer https://github.com/kubernetes/website/pull/39914 with my suggestions
